third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp

```diff
@@ -772,6 +843,389 @@ LogicalResult Pingponger::sliceDot(OpBuilder &builder, Location loc,
+
+    auto warpsPerCTA = encoding.getWarpsPerCTA();
+    auto sizePerThread = encoding.getSizePerThread();
+    SmallVector<unsigned> threadPerWarp(warpsPerCTA.size(), 0);
+    for (size_t dim = 0; dim < numDims; ++dim) {
+      threadPerWarp[dim] =
+          slicedShape[dim] / (warpsPerCTA[dim] * sizePerThread[dim]);
+    }
+    assert(mlir::product(threadPerWarp) == 64);
+
+    auto newEncoding = ttg::BlockedEncodingAttr::get(
+        ctx, sizePerThread, threadPerWarp, warpsPerCTA, encoding.getOrder(),
+        encoding.getCTALayout());
+
+    auto converTensor = [&](mlir::TypedValue<mlir::RankedTensorType> tensor) {
+      RankedTensorType newType = nullptr;
+      Value newTensor = nullptr;
+      RankedTensorType slicedTensorType = nullptr;
+      if (tensor) {
+        assert(encoding == tensor.getType().getEncoding());
+        auto elemType = tensor.getType().getElementType();
+        newType = RankedTensorType::get(origShape, elemType, newEncoding);
+        newTensor =
+            builder
+                .create<ttg::ConvertLayoutOp>(tensor.getLoc(), newType, tensor)
+                .getResult();
+        slicedTensorType =
+            RankedTensorType::get(slicedShape, elemType, newEncoding);
+      }
+
+      return std::make_tuple(newType, newTensor, slicedTensorType);
+    };
+
+    mlir::TypedValue<mlir::RankedTensorType> origMask = nullptr;
+    mlir::TypedValue<mlir::RankedTensorType> origOtherTensor = nullptr;
+
+    if (auto value = asyncCopy.getMask()) {
+      origMask = dyn_cast<decltype(origMask)>(value);
+    }
+    if (auto value = asyncCopy.getOther()) {
+      origOtherTensor = cast<decltype(origOtherTensor)>(value);
+    }
+
+    auto [newSrcType, newSrcPointers, slicedSrcType] =
+        converTensor(srcPointers);
+    auto [newMaskType, newMask, slicedMaskType] = converTensor(origMask);
+    auto [newOtherType, newOther, slicedOtherType] =
+        converTensor(origOtherTensor);
+
+    auto extract = [&builder](Type resType, Value src,
+                              DenseI64ArrayAttr &offset) {
+      Value resValue = nullptr;
+      if (src) {
+        resValue = builder.create<triton::amdgpu::ExtractSliceOp>(
+            src.getLoc(), resType, src, offset);
+      }
+      return resValue;
+    };
+
+    assert(slicedDim != -1);
+    SmallVector<Value> newCommits;
+    auto numReps = origShape[slicedDim] / slicedShape[slicedDim];
+    for (size_t rep = 0; rep < numReps; ++rep) {
+      SmallVector<int64_t> offset(slicedShape.size(), 0);
+      offset[slicedDim] = slicedShape[slicedDim] * rep;
+      auto offsetAttr = DenseI64ArrayAttr::get(ctx, offset);
+
+      auto extractedSrc = extract(slicedSrcType, newSrcPointers, offsetAttr);
+      auto extractedMask = extract(slicedMaskType, newMask, offsetAttr);
+      auto extractedOther = extract(slicedOtherType, newOther, offsetAttr);
+
+      auto newAsyncCopy = builder.create<ttg::AsyncCopyGlobalToLocalOp>(
```

Comment:
Can we keep pointers to the new copy slices as well?
We may want to relocate everything.

Reply:
Added attributes to `newAsyncCopy` - i.e., `DotOpIdx` and `sliceId`